### DISCUSSION TITLE
Major improvements across app

### DISCUSSION THREAD

**Discussion by LogicLock** [pieces-app/support#28] [Ideas]
### Introduction

I have been searching for some time for the perfect snippet program for my needs. I have some big visions laid out that I would like to start implementing, and the program is part of the requirements. Inquiring about the available features on discord, I suggested the idea of "collections" and, on the developer's advice, began to dissect ideas and possible improvements to the application. I will divide the ideas into categories and more or less describe the process of action. English is not my native language and I speak it at an intermediate level so I apologize in advance for any mistakes.

### Main View
![2024-01-12 at 00 51 57](https://github.com/pieces-app/support/assets/147541302/735d3267-4053-4f3e-8a37-05d61faf0aec)

The home screen and available views are, in my opinion, what is most important when it comes to using the program. 

Let's assume such a scenario:
I am a developer using at least 2 programming languages + some additional scripts like powershell or Linux + git. Such a set up is as likely as possible already for a mid developer, who's level of knowledge is sufficiently discernible to use a lot of programs that help on a daily basis in the code, like snippet manager. In this case, it is very likely that the developer will create at least 50 snippets in different languages.

In my opinion, none of the views in this list is good for managing and handling a very large number of snippets. For someone who is a visual learner and does not remember everything, every view from the available options is bad. Don't get me wrong, these are nice options, but in the above case, not always good and proven.

**First improvement - Collections**
Probably each of you has used folders, subfolders, notebooks storing "books" and in it many different thematic notes. 
My idea to improve the view is to be able to create a collection. The collection would be something like a folder. Each collection would consist of a default or custom icon, title, description, tags and possibly additional links as a separate section. Each collection allows you to store x snippets. 
Example: We are using the C# language and want to create some basic snippets in this language. I create a collection called "C# Basic" and in it I add some snippets. In a view such as List View, I see this collection, expand it and I have all the snippets in this collection. This will certainly help in managing what we currently have. If someone wants to create snippets for UI/UX, create some algorithms or synchronization mechanisms, they can sort it nicely by topic. Here you might be tempted to add the ability to automatically add tags. If for a collection we have added the tag "c# 2.0" and "c# basic" then in my opinion it should be added automatically to each snippet, but such option can be turned off.
In my opinion, such a collection should be able to be made available in its entirety, sharing from link or different methods.

**Second improvement - Workspaces**
I was also thinking about something like WorkSpaces in two possibilities.
The first possibility is to create workspaces as separate environments e.g. for work, for home use, etc. Each environment is a separate entity that stores snippets, but one between the other knows nothing about each other.

The second possibility is to create workspaces to separate different programming languages. They would not be separate entities, because they would have knowledge about each other (because, for example, I will create the same code in two languages), but if we created a workspace named "C#" then it should hold snippets only in that language. Here rather loosely thinking and a matter of discussion

**Third improvement - graph or links**
I was thinking a lot about something like being able to combine dependencies or requirements. Let's say I'm creating code, but it's so large that it needs to be broken into several files and to make it look readable I create several snippets in parts. I should be able to combine the dependencies and requirements that 1 part is required to realize 2 and 1 and 2 parts depend on creating the form. This may not be the best example, but there are situations when one piece depends on another or they are somehow connected together. I do not know if I would make a separate view for this available from the level of e.g. collections or workspace or create such a functionality so that you can do it from the level of the program.

### Tags
Tags shouldn't just be tags but should be the basis for linked structures and links. 

**First improvement - global tags**
In my opinion, it should be possible to manage your own global tags. At the moment I create them by typing them in manually every time inside snippet, but I won't remember them all word for word. Within a workspace or account, I should be able to create tags that I will reuse in snippets. This will mainly give control over a uniform form and confidence in the links. They will be displayed and available to be added in each newly created snippet.

**Second improvement - linking tags**
Tags should have linking features and should tie snippets together, for example, by feature or language. With such functionality, it should be possible to filter by tags, for example, in List View. Also it will help in **graph or link view**.

I have other ideas, but I'm throwing in the real and latest ones. Maybe they will find others and even better ones. For now, I'm starting the discussion and as soon as something hits I'll be dropping it off

**Comment by eumel1968** [comment]
That would be awesome!
